tests/compilation_failure_tests/CMakeLists.txt

```diff
@@ -0,0 +1,38 @@
+#
+# This directory contains files aimed to verify that constructs that
+# are supposed to fail at compile time, indeed do so.
+# To prevent bit rot, the same source file is compiled twice with
+# the macro COMPILATION_TEST_USE_FAILING_CODE set to 0 or 1.
+#
+
+# internal function for add_dual_compile_test
+# which adds a target and a test target trying to compile it
+function(detail_add_dual_compile_test TEST_NAME TEST_FILE)
+  add_executable(${TEST_NAME} ${TEST_FILE})
+  set_target_properties(${TEST_NAME} PROPERTIES
+                        EXCLUDE_FROM_ALL TRUE
+                        EXCLUDE_FROM_DEFAULT_BUILD TRUE)
+  add_test(NAME ${TEST_NAME}
+           COMMAND ${CMAKE_COMMAND} --build . --target ${TEST_NAME} --config $<CONFIGURATION>
+           WORKING_DIRECTORY ${CMAKE_BINARY_DIR})
+endfunction(add_compile_test)
+
+# adds a compilation test. two targets are created, one expected to
```

Comment:
I don't this this objection should block checkin, but it does seem like it'd be a little easier to work with this directory if only failing compilation tests were here. We have success tests elsewhere, which will / should test the result as well, where compile tests can only test compilation.

Reply:
I agree with you comment, but I value keeping the code from bit rot higher! The positive compile test is just to make sure the test code still makes sense. That way, it will be seen during refactoring and it is more difficult to forget to update also the code that is supposed not to compile.

I think the success tests, like the code examples, are great to have as a part of the normal build. Keeping the documentation correct is really nice!

Reply:
That's fair :)